compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -212,14 +212,19 @@ object PatternMatcher {
     private def patternPlan(scrutinee: Symbol, tree: Tree, onSuccess: Plan, onFailure: Plan): Plan = {
 
       /** Plan for matching `selectors` against argument patterns `args` */
-      def matchArgsPlan(selectors: List[Tree], args: List[Tree], onSuccess: Plan): Plan =
-        args match {
-          case arg :: args1 =>
-            val selector :: selectors1 = selectors
-            letAbstract(selector)(
-              patternPlan(_, arg, matchArgsPlan(selectors1, args1, onSuccess), onFailure))
-          case Nil => onSuccess
-        }
+      def matchArgsPlan(selectors: List[Tree], args: List[Tree], onSuccess: Plan): Plan = {
+        def matchArgsSelectorsPlan(selectors: List[Tree], syms: List[Symbol]): Plan =
+          selectors match {
+            case selector :: selectors1 => letAbstract(selector)(sym => matchArgsSelectorsPlan(selectors1, sym :: syms))
+            case Nil => matchArgsPatternPlan(args, syms.reverse)
+          }
+        def matchArgsPatternPlan(args: List[Tree], syms: List[Symbol]): Plan =
+          ((args, syms): @unchecked) match { // both lists should always have the same size
+            case (arg :: args1, sym :: syms1) => patternPlan(sym, arg, matchArgsPatternPlan(args1, syms1), onFailure)
+            case (Nil, Nil) => onSuccess
```

Comment:
I guess you can speed this up by doing like it was done initially:
```scala
args match {
  case arg :: args1 =>
    val sym :: sym = syms
    patternPlan(sym, arg, matchArgsPatternPlan(args1, syms1), onFailure)
  case Nil => onSuccess
}
```
You don't need to construct and deconstruct a tuple anymore

Reply:
I would need to add an assertion in the `Nil` case to ensure failure. At some point, we should make that optimization work, there is no good reason for it to create the in those cases.

Reply:
Actually with `-optimize` we already remove the deconstruction of the tuple. It currently becomes:
```scala
        case val x1: Tuple2 = new Tuple2(xs, ys)
        if 
          xs.isInstanceOf[scala.collection.immutable.::].&&(
            ys.isInstanceOf[scala.collection.immutable.::]
          )
         then 
          patternPlan(sym, arg, matchArgsPatternPlan(args1, syms1), onFailure)
         else 
          {
             def case2(case x16: scala.collection.immutable.List): String = 
              if Nil().==(x16).&&(Nil().==(ys)) then onSuccess else 
                {
                   def case1(): String = throw new MatchError(x1)
                  case1()
                }
            case2(xs)
          }
```

There is still a spurious  `new Tuple2`.

Reply:
Even if the `-optimise` option removed it, this is not something we want to enable by default since it adds a significant compilation overhead